lib/bloc_record/persistence.rb

```diff
@@ -22,6 +22,85 @@ def create(attrs)
       new(data)
     end
 
+    def update(ids, updates)
+      # add a check for class of updates in case an array of attributes for diff
+      # records is passed (assignment 5)
+      if updates.class == Hash
+        updates = BlocRecord::Utility.convert_keys(updates)
+        updates.delete("id")
+        updates_array = updates.map { |key, value| "#{key}=#{BlocRecord::Utility.sql_strings(value)}" }
+      elsif updates.class == Array
+        updates_array = updates
+      end
+
+      if ids.class == Fixnum
+        where_clause = "WHERE id = #{ids};"
+      elsif ids.class == Array
+        where_clause = ids.empty? ? ";" : "WHERE id IN (#{ids.join(",")});"
+      else
+        where_clause = ";"
+      end
+
+      connection.execute <<-SQL
+        UPDATE #{table}
+        SET #{updates_array * ","} #{where_clause}
+      SQL
+
+      true
+    end
+
+    def destroy(*id)
+      if id.length > 1
+        where_clause = "WHERE id IN (#{id.join(",")})"
+      else
+        where_clause = "WHERE id = #{id.first};"
+      end
+
+      connection.execute <<-SQL
+        DELETE FROM #{table}
+        #{where_clause}
+      SQL
+
+      true
+    end
+
+    def destroy_all(*conditions=nil)
+      # #3 from assignment, using execute(sql, params) syntax
+      if conditions.length > 1
+        sql_expression = conditions.shift
+        params = conditions
+
+        sql = <<-SQL
+          DELETE FROM #{table}
+          WHERE #{sql_expression}
+        SQL
+
+        connection.execute(sql, params)
+      # Hash passed in a condition
+      elsif conditions.first.class == Hash && !conditions.first.empty?
+        conditions = BlocRecord::Utility.convert_keys(conditions.first)
+        conditions = conditions.map {|key, value| "#{key} = #{BlocRecord::Utility.sql_strings(value)}"}.join(" and ")
+
+        connection.execute <<-SQL
+          DELETE FROM #{table}
+          WHERE #{conditions}
+        SQL
+      # #2 from assignment. String passed as condition
+      elsif conditions.first.class == String && !conditions.first.empty?
+        connection.execute <<-SQL
+          DELETE FROM #{table}
+          WHERE #{conditions.first}
+        SQL
+      else
```

Comment:
This is not specified whatsoever, but I thought I would bring it up as a thought experiment.

Should passing in an empty string or empty `Hash` result in everything getting deleted? In your implementation, that is what happens. I can see both ways, really. There are no conditions, so no conditions should be applied. Or set of conditions were provided and it happens to be empty, meaning nothing can match it.

The latter feels weaker to me semantically. However, I could imagine if someone was building the conditions in code and it just so happened none of the logic applied, it may surprise them to see their database got nuked. It is a semantics versus functional use case argument, which I personally hate to fall in to.